### PR TITLE
#5469 Add a lock to avoid threads competition on touching stopping cluster

### PR DESCRIPTION
**Pull request by chia7712**
### Checklist

- [ ] [Review Contribution Guidelines](https://oharastream.github.io/en/docs/master/contributing/)
- [ ] Add at least one reviewer (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [ ] Add an assignee (of course, you are the assignee by default)
- [ ] Do not set milestone, projects and labels (PR labels are added automatically by Jenkins)
- [ ] Link to an issue
- [ ] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (If the failed tests are unrelated to this PR, add a comment in the PR thread)

### CODE REVIEW

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/route/package.scala [hunk 1]
```diff
@@ -51,6 +54,37 @@ package object route {
...
+    * is unable to accept connection. However, our worker route has to check the active connectors before stopping it.
+    * Hence, the later request of stopping worker hangs since it can't get connectors information from stopping worker.
+    * If our UI closes connection (due to timeout), the connection (on server) ends up being a ghost and it will stop worker
+    * cluster again (randomly!!!)
+    */
+  private[this] def hookOfStopAction[Cluster <: ClusterInfo](
+    hookBeforeStop: HookOfAction[Cluster]
+  )(implicit collie: Collie, executionContext: ExecutionContext) = new HookOfAction[Cluster] {
+    override def apply(clusterInfo: Cluster, subName: String, params: Map[String, String]): Future[Unit] = {
+      if (STOP_LOCK.tryLock()) {
```
Comment (saivirtue) [inline ohara-configurator/src/main/scala/oharastream/ohara/configurator/route/package.scala:70]: 請問一下，當引入lock機制後，是否同時間只能有一個"StopAction"被觸發?

**Comment by jackyoh** [comment]
retry

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/route/package.scala (hunk 1 above)
Comment (chia7712) [reply ohara-configurator/src/main/scala/oharastream/ohara/configurator/route/package.scala:70]: 沒錯

先看看這樣簡單的方法是否可行

另一個解法是不同的key用不同的lock，會比較複雜但是比較平行

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/route/package.scala [hunk 2]
```diff
@@ -51,6 +54,37 @@ package object route {
...
+        val f = hookBeforeStop(clusterInfo, subName, params)
+        f.onComplete {
+          case Success(_) => // we will unlock later
```
Comment (jackyoh) [inline ohara-configurator/src/main/scala/oharastream/ohara/configurator/route/package.scala:73]: > // we will unlock later

要在 0.11.x 版做，還是下個版本再做

Comment (chia7712) [reply ohara-configurator/src/main/scala/oharastream/ohara/configurator/route/package.scala:73]: 這段指得是下一個future就會去unlock

Comment (jackyoh) [reply ohara-configurator/src/main/scala/oharastream/ohara/configurator/route/package.scala:76]: OK, I got it.

**Comment by jackyoh** [comment]
retry

File: ohara-configurator/src/main/scala/oharastream/ohara/configurator/route/package.scala [hunk 3]
```diff
@@ -51,6 +55,39 @@ package object route {
   private[route] val PAUSE_COMMAND: String   = oharastream.ohara.client.configurator.PAUSE_COMMAND
   private[route] val RESUME_COMMAND: String  = oharastream.ohara.client.configurator.RESUME_COMMAND
 
+  private[this] val STOPPING_FLAGS = new ConcurrentHashMap[KeyKind, AtomicBoolean]()
```
Comment (chia7712) [inline ohara-configurator/src/main/scala/oharastream/ohara/configurator/route/package.scala:58]: 我將lock改成用atomic boolean以免遇到不同執行緒去解除lock的意外

Comment (jackyoh) [reply ohara-configurator/src/main/scala/oharastream/ohara/configurator/route/package.scala:58]: 確保執行緒安全

**Comment by chia7712** [comment]
retry collieIT

**Comment by chia7712** [comment]
retry connectorIT

**Comment by chia7712** [comment]
retry streamIT

**Pull request merged**